sky/resources.py

```diff
@@ -275,33 +275,47 @@ def _set_region_zone(self, region: Optional[str],
         self._region, self._zone = self._cloud.validate_region_zone(
             region, zone)
 
-    def get_valid_region_zones(self) -> List[Tuple[clouds.Region, clouds.Zone]]:
+    def get_valid_region_zones_for_launchable(
+            self) -> List[Tuple[clouds.Region, List[clouds.Zone]]]:
         """Returns a list of (region, zone) that can provision this Resources"""
         assert self.is_launchable()
 
+        # Get resource-offering regions and zones.
+        region_zones = list(
+            self._cloud.region_zones_provision_loop(
+                instance_type=self._instance_type,
+                accelerators=self.accelerators,
+                use_spot=self._use_spot))
         if isinstance(self._cloud, clouds.GCP):
-            gcp_region_zones = list(
-                self._cloud.region_zones_provision_loop(
-                    instance_type=self._instance_type,
-                    accelerators=self.accelerators,
-                    use_spot=self._use_spot))
-
-            # GCP provision loop yields 1 zone per request.
-            # For consistency with other clouds, we group the zones in the
-            # same region.
-            region_zones = []
+            # GCP provision loop yields 1 zone per request. For consistency with
+            # other clouds, we group the zones in the same region.
+            # This is required by Optimizer.
```

Comment:
Took me a while to see why this is required by 

https://github.com/skypilot-org/skypilot/blob/1eeecee1d1fde6d5521de8cb27ef5278df303b57/sky/optimizer.py#L861

I think the reason is, otherwise the launchables there would duplicate a region. 

How about we enforce this in the interface, by returning say `Set[Region]` where each Region has the appropriate Zones added? Or something like `Dict[str, Set[str]]` (region name to a set of zone names)?

Reply:
Sorry I didn't get your point here. What do you mean by the "interface" here? And how is your data structure different from `List[Tuple[Region, List[Zone]]]` returned by this function?

Reply:
(No action needed) The current interface is 
```
get_valid_region_zones_for_launchable() -> List[Tuple[clouds.Region, List[clouds.Zone]]]
```
A caller looks at this interface and won't know whether the returned list contains 1 tuple per region. (It's likely, but not clear.) So the caller needs to look at the docstr/impl to see if this is the case.

If the interface were 

```
get_valid_region_zones_for_launchable() -> Dict[RegionName, Set[ZoneName]]
```
or something like that, then the property is inferable from the signature of this func.

---

I think the above is minor. More relevant is add to L292 the reasons that the code below is needed by the optimizer.

Reply:
Got it. Added a comment that this is to not duplicate the regions.